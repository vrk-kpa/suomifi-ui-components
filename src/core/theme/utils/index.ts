export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
