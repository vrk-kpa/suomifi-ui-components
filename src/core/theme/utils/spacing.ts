import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (props: SpacingProps | undefined): string => {
  if (!props) return '';

  const cssStyles = Object.entries(props)
    .map(([key, value]) =>
      getCSSSpacing(defaultSuomifiTheme, key as keyof SpacingProps, value),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return `margin-right: ${amount}; margin-left: ${amount};`;
    case 'my':
      return `margin-top: ${amount}; margin-bottom: ${amount};`;
    case 'px':
      return `padding-right: ${amount}; padding-left: ${amount};`;
    case 'py':
      return `padding-top: ${amount}; padding-bottom: ${amount};`;
    default:
      return `${[cssSelector[key]]}: ${amount};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
