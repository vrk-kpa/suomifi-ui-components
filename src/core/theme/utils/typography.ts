import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
