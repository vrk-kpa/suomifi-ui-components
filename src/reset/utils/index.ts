import { css } from 'styled-components';

const normalize = {
  html:
    'line-height: 1.15; -ms-text-size-adjust: 100%; -webkit-text-size-adjust: 100%;',
  h1: 'font-size: 2em; margin: 0.67em 0;',
  'dl dl': 'margin: 0;',
  'dl ol': 'margin: 0;',
  'dl ul': 'margin: 0;',
  'ol dl': 'margin: 0;',
  'ul dl': 'margin: 0;',
  'ol ol': 'margin: 0;',
  'ol ul': 'margin: 0;',
  'ul ol': 'margin: 0;',
  'ul ul': 'margin: 0;',
  hr: 'box-sizing: content-box; height: 0; overflow: visible;',
  main: 'display: block;',
  pre: 'font-family: monospace, monospace; font-size: 1em;',
  a: 'background-color: transparent;',
  'abbr[title]':
    'text-decoration: underline; text-decoration: underline dotted;',
  b: 'font-weight: bolder;',
  strong: 'font-weight: bolder;',
  code: 'font-family: monospace, monospace; font-size: 1em;',
  kbd: 'font-family: monospace, monospace; font-size: 1em;',
  samp: 'font-family: monospace, monospace; font-size: 1em;',
  small: 'font-size: 80%;',
  audio: 'display: inline-block;',
  video: 'display: inline-block;',
  'audio:not([controls])': 'display: none; height: 0;',
  img: 'border-style: none;',
  'svg:not(:root)': 'overflow: hidden;',
  button:
    'margin: 0;overflow: visible; text-transform: none;-webkit-appearance: button;',
  input: 'margin: 0;overflow: visible;',
  select: 'margin: 0;text-transform: none;',
  '[type="button"]': '-webkit-appearance: button;',
  '[type="reset"]': '-webkit-appearance: button;',
  '[type="submit"]': '-webkit-appearance: button;',
  fieldset: 'padding: 0.35em 0.75em 0.625em;',
  legend:
    'box-sizing: border-box; color: inherit; display: table; max-width: 100%; white-space: normal;',
  progress: 'display: inline-block; vertical-align: baseline;',
  textarea: 'margin: 0; overflow: auto;',
  '[type="checkbox"]': 'box-sizing: border-box; padding: 0;',
  '[type="radio"]': 'box-sizing: border-box; padding: 0;',
  '[type="search"]': '-webkit-appearance: textfield; outline-offset: -2px;',
  '::-webkit-inner-spin-button': 'height: auto;',
  '::-webkit-outer-spin-button': 'height: auto;',
  '::-webkit-input-placeholder': 'color: inherit; opacity: 0.54;',
  '::-webkit-search-decoration': '-webkit-appearance: none;',
  '::-webkit-file-upload-button': '-webkit-appearance: button; font: inherit;',
  '::-moz-focus-inner': 'border-style: none; padding: 0;',
  ':-moz-focusring': 'outline: 1px dotted ButtonText;',
  ':-moz-ui-invalid': 'box-shadow: none;',
  details: 'display: block;',
  dialog:
    // eslint-disable-next-line max-len
    'background-color: white; border: solid; color: black; display: block; height: -moz-fit-content; height: -webkit-fit-content; height: fit-content; left: 0; margin: auto; padding: 1em; position: absolute; right: 0; width: -moz-fit-content; width: -webkit-fit-content; width: fit-content;',
  'dialog:not([open])': 'display: none;',
  summary: 'display: list-item;',
  canvas: 'display: inline-block;',
  template: 'display: none;',
  '[hidden]': 'display: none;',
} as Record<string, string>;

const common = css`
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;

  ::before,
  ::after {
    box-sizing: border-box;
  }
`;

const resets = {
  normalize,
  common,
};

const selectorDeclaration = (selector: string, styles: string) =>
  `&${selector} {
    ${styles}
  }`;

const selectorsAndCss = (selector: string) => {
  const styles = !!normalize && normalize[selector];
  return !!styles ? selectorDeclaration(selector, styles) : '';
};

/**
 * Return selectors with resets in Emotion-css
 * @param selectors string or array of strings
 */
const resetWithSelectors = (selectors: string | string[]): string => {
  if (Array.isArray(selectors) && selectors.length > 0) {
    return `
      ${selectors.map((s) => selectorsAndCss(s)).join('')}
    `;
  }
  if (typeof selectors === 'string') {
    return selectorsAndCss(selectors);
  }
  return '';
};

export { resets, resetWithSelectors };

// TODO: Add tests
// TODO: create versioned interface for normalize.css
