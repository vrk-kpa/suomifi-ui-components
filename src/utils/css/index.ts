export { alphaHex, hslaToHex } from './colors';
export { allStates, absolute } from './pseudo';
export {
  clearfix,
  cssObjectToCss,
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from './utils';
