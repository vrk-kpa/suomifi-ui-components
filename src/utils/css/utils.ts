import { css, FlattenSimpleInterpolation } from 'styled-components';
import { ValueUnit } from 'suomifi-design-tokens';
export { FlattenSimpleInterpolation };

export const clearfix = css`
  &:after {
    display: block;
    content: '';
    clear: both;
  }
`;

/**
 * Return FlattenSimpleInterpolation compatible string or number
 * @param cssValue compatible value or {value, unit} pair-object
 */
const cssValueToString = (cssValue: number | string | ValueAndUnit) => {
  if (!!cssValue && typeof cssValue === 'object' && 'value' in cssValue) {
    const { value, unit } = cssValue;
    return !!unit ? `${value}${unit}` : value;
  }
  return cssValue;
};

const camelToSnake = (string: string) =>
  string.replace(/[\w]([A-Z])/g, m => `${m[0]}-${m[1]}`).toLowerCase();
/**
 * Convert CSSObject to CSS FlattenSimpleInterpolation
 * @param value CSSObject
 */
export const cssObjectToCss = <T>(value: T) => css`
  ${Object.entries(value)
    .map(([key, value]) => `${camelToSnake(key)}: ${cssValueToString(value)};`)
    .join('')}
`;

/**
 * Convert CSSObject tokens to CSS FlattenSimpleInterpolation
 * @param tokens Tokens of CSSObjects
 */
export const cssObjectsToCss = <T, K extends keyof T>(tokens: T) =>
  Object.entries(tokens).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: cssObjectToCss(value),
    }),
    {} as { [key in K]: FlattenSimpleInterpolation },
  );
