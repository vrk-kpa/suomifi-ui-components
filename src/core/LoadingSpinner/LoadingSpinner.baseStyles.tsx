import { css } from 'styled-components';
import { SuomifiTheme } from '../theme';
import { font, element } from '../theme/reset';
import { MarginProps, buildSpacingCSS } from '../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${element(theme)}
  ${font(theme)('bodyTextSmall')}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  line-height: 1;
  &.fi-loadingSpinner {
    display: block;
    text-align: center;
    & .fi-loadingSpinner_icon {
      display: inline-block;
      width: 40px;
      height: 40px;
    }
    & .fi-loadingSpinner_text {
      font-size: 18px;
      font-weight: 600;
      text-align: center;
      margin-top: 6px;
    }
    &.fi-loadingSpinner-textAlign--right {
      display: flex;
      align-items: center;

      & .fi-loadingSpinner_text {
        margin-left: 10px;
        margin-top: 0;
        line-height: 1.5em;
      }
    }
    &.fi-loadingSpinner--small {
      .fi-loadingSpinner_icon {
        width: 24px;
        height: 24px;
      }
      & .fi-loadingSpinner_text {
        font-size: 16px;
        font-weight: normal;
        margin-top: 0;
      }
    }
    &.fi-loadingSpinner--loading {
      & .fi-loadingSpinner_icon {
        animation: rotation 1.5s infinite linear;
      }
    }
    &.fi-loadingSpinner--success {
      .fi-icon-base-fill {
        fill: ${theme.colors.successBase};
      }
    }
    &.fi-loadingSpinner--failed {
      .fi-icon-base-fill {
        fill: ${theme.colors.alertBase};
      }
    }
  }
  @keyframes rotation {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(359deg);
    }
  }
  @media (prefers-reduced-motion) {
    &.fi-loadingSpinner.fi-loadingSpinner--loading {
      .fi-loadingSpinner_icon {
        animation: rotation 10s infinite linear;
      }
    }
  }
`;
