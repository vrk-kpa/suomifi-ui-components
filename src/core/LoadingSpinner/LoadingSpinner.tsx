import React, { Component, forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { baseStyles } from './LoadingSpinner.baseStyles';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../theme/utils/spacing';
import { HtmlDiv, HtmlDivProps, HtmlDivWithRef } from '../../reset';
import { VisuallyHidden } from '../VisuallyHidden/VisuallyHidden';
import {
  IconCheckCircleFilled,
  IconErrorFilled,
  IconPreloader,
} from 'suomifi-icons';
import { filterDuplicateKeys } from '../../utils/common/common';

export type LoadingSpinnerStatus = 'loading' | 'success' | 'failed';
export interface LoadingSpinnerProps extends MarginProps, HtmlDivProps {
  /** CSS class for custom styles */
  className?: string;
  /** HTML id attribute */
  id?: string;
  /** Status text shown next to the loading icon */
  text: string;
  /** Aligns the text on the bottom or on the right side of the animation svg
   * @default 'bottom'
   */
  textAlign?: 'bottom' | 'right';
  /** Shows or hides the status text
   * @default 'visible'
   */
  textVisibility?: 'visible' | 'hidden';
  /**
   * `'loading'` | `'success'` | `'failed'`;
   *
   * Status that the component indicates. Affects the icon used.
   * @default 'loading'
   */
  status?: LoadingSpinnerStatus;
  /** Component variant. Affects the size of the icon and the text
   * @default 'normal'
   */
  variant?: 'normal' | 'small';
  /** Ref is placed to the outermost div element of the component. Alternative to React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLDivElement>;
}

const baseClassName = 'fi-loadingSpinner';

export const loadingSpinnerClassNames = {
  text: `${baseClassName}_text`,
  textAlignBottom: `${baseClassName}-textAlign--bottom`,
  textAlignRight: `${baseClassName}-textAlign--right`,
  small: `${baseClassName}--small`,
  failed: `${baseClassName}--failed`,
  loading: `${baseClassName}--loading`,
  success: `${baseClassName}--success`,
  icon: `${baseClassName}_icon`,
};

class BaseLoadingSpinner extends Component<LoadingSpinnerProps> {
  render() {
    const {
      className,
      id,
      text,
      textAlign = 'bottom',
      textVisibility = 'visible',
      variant = 'normal',
      status = 'loading',
      style,
      forwardedRef,
      ...rest
    } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);

    return (
      <HtmlDivWithRef
        className={classnames(baseClassName, className, {
          [loadingSpinnerClassNames.loading]: status === 'loading',
          [loadingSpinnerClassNames.failed]: status === 'failed',
          [loadingSpinnerClassNames.success]: status === 'success',
          [loadingSpinnerClassNames.small]: variant === 'small',
          [loadingSpinnerClassNames.textAlignRight]: textAlign === 'right',
        })}
        as="section"
        id={id}
        ref={forwardedRef}
        {...passProps}
        style={style}
      >
        {status === 'loading' && (
          <IconPreloader className={loadingSpinnerClassNames.icon} />
        )}
        {status === 'success' && (
          <IconCheckCircleFilled className={loadingSpinnerClassNames.icon} />
        )}
        {status === 'failed' && (
          <IconErrorFilled className={loadingSpinnerClassNames.icon} />
        )}

        {textVisibility === 'visible' ? (
          <HtmlDiv className={loadingSpinnerClassNames.text}>{text}</HtmlDiv>
        ) : (
          <VisuallyHidden>{text}</VisuallyHidden>
        )}
      </HtmlDivWithRef>
    );
  }
}
const StyledLoadingSpinner = styled(
  (props: LoadingSpinnerProps & SuomifiThemeProp & GlobalMarginProps) => {
    const { theme, globalMargins, ...passProps } = props;
    return <BaseLoadingSpinner {...passProps} />;
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.loadingSpinner,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;
const LoadingSpinner = forwardRef(
  (props: LoadingSpinnerProps, ref: React.Ref<HTMLDivElement>) => {
    const { ...passProps } = props;
    return (
      <SpacingConsumer>
        {({ margins }) => (
          <SuomifiThemeConsumer>
            {({ suomifiTheme }) => (
              <StyledLoadingSpinner
                forwardedRef={ref}
                theme={suomifiTheme}
                globalMargins={margins}
                {...passProps}
              />
            )}
          </SuomifiThemeConsumer>
        )}
      </SpacingConsumer>
    );
  },
);

LoadingSpinner.displayName = 'LoadingSpinner';
export { LoadingSpinner };
