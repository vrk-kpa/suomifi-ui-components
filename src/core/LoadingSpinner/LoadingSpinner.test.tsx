import React from 'react';
import { render } from '@testing-library/react';
import { LoadingSpinner } from './LoadingSpinner';
import { axeTest } from '../../utils/test';

describe('props', () => {
  describe('textVisibility should be visually-hidden', () => {
    const { getByText } = render(
      <LoadingSpinner text="To be hidden" textVisibility="hidden" />,
    );
    const spinner = getByText('To be hidden');
    expect(spinner).toHaveClass('fi-visually-hidden');
  });

  describe('id', () => {
    const LoadingSpinnerWithElementId = (
      <LoadingSpinner id="TestId" text="Loading data" />
    );
    const { container } = render(LoadingSpinnerWithElementId);
    expect(container.querySelector('#TestId')).toHaveClass('fi-loadingSpinner');
  });

  describe('textAlign', () => {
    it('should apply styles based on given alignMent', () => {
      const textAlignLoadingSpinner = (
        <LoadingSpinner textAlign="right" text="Loading data" />
      );
      const { container } = render(textAlignLoadingSpinner);
      expect(container.firstChild).toHaveClass(
        'fi-loadingSpinner-textAlign--right',
      );
    });
  });

  describe('text', () => {
    it('should have given text content', () => {
      const TextLoadingSpinner = (
        <LoadingSpinner text="Loading data" className="custom-class" />
      );
      const { getByText } = render(TextLoadingSpinner);
      const text = getByText('Loading data');
      expect(text).toHaveClass('fi-loadingSpinner_text');
    });
  });

  describe('className', () => {
    const customClassLoadingSpinner = (
      <LoadingSpinner text="Loading data" className="custom-class" />
    );

    it('contains base classname', () => {
      const { container } = render(customClassLoadingSpinner);
      expect(container.firstChild).toHaveClass('fi-loadingSpinner');
    });

    it('contains custom classname', () => {
      const { container } = render(customClassLoadingSpinner);
      expect(container.firstChild).toHaveClass('custom-class');
    });
  });

  describe('margin', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(<LoadingSpinner text="Test" margin="xs" />);
      expect(container.firstChild).toHaveStyle('margin: 10px');
    });

    it('should have margin style overridden by style prop', async () => {
      const { container } = render(
        <LoadingSpinner text="Test" margin="xs" style={{ margin: 2 }} />,
      );
      expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
    });
  });

  describe('snapshot', () => {
    const TestLoadingSpinner = <LoadingSpinner text="Loading data" />;
    it('should match snapshot', () => {
      const { container } = render(TestLoadingSpinner);
      expect(container).toMatchSnapshot();
    });
  });

  describe('accessibility', () => {
    it('should not have any basuc accessibility issues', () => {
      const TestLoadingSpinner = <LoadingSpinner text="Loading data" />;
      axeTest(TestLoadingSpinner);
    });
  });
});
