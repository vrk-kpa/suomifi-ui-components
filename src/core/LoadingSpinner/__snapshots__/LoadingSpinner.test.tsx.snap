// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props snapshot should match snapshot 1`] = `
.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c3 ::before,
.c3 ::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c2 {
  vertical-align: baseline;
}

.c2.fi-static-icon {
  display: inline-block;
}

.c2.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c2.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c1 {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  line-height: 1;
}

.c1.fi-loadingSpinner {
  display: block;
  text-align: center;
}

.c1.fi-loadingSpinner .fi-loadingSpinner_icon {
  display: inline-block;
  width: 40px;
  height: 40px;
}

.c1.fi-loadingSpinner .fi-loadingSpinner_text {
  font-size: 18px;
  font-weight: 600;
  text-align: center;
  margin-top: 6px;
}

.c1.fi-loadingSpinner.fi-loadingSpinner-textAlign--right {
  display: flex;
  align-items: center;
}

.c1.fi-loadingSpinner.fi-loadingSpinner-textAlign--right .fi-loadingSpinner_text {
  margin-left: 10px;
  margin-top: 0;
  line-height: 1.5em;
}

.c1.fi-loadingSpinner.fi-loadingSpinner--small .fi-loadingSpinner_icon {
  width: 24px;
  height: 24px;
}

.c1.fi-loadingSpinner.fi-loadingSpinner--small .fi-loadingSpinner_text {
  font-size: 16px;
  font-weight: normal;
  margin-top: 0;
}

.c1.fi-loadingSpinner.fi-loadingSpinner--loading .fi-loadingSpinner_icon {
  animation: rotation 1.5s infinite linear;
}

.c1.fi-loadingSpinner.fi-loadingSpinner--success .fi-icon-base-fill {
  fill: hsl(166, 90%, 34%);
}

.c1.fi-loadingSpinner.fi-loadingSpinner--failed .fi-icon-base-fill {
  fill: hsl(3, 59%, 43%);
}

@media (prefers-reduced-motion) {
  .c1.fi-loadingSpinner.fi-loadingSpinner--loading .fi-loadingSpinner_icon {
    animation: rotation 10s infinite linear;
  }
}

<div>
  <section
    class="c0 fi-loadingSpinner c1 fi-loadingSpinner--loading"
  >
    <svg
      aria-hidden="true"
      class="fi-icon c2 fi-loadingSpinner_icon"
      focusable="false"
      height="1em"
      viewBox="0 0 40 40"
      width="1em"
    >
      <path
        class="fi-icon-component-brand-fill"
        d="M20 0v3C10.611 3 3 10.611 3 20s7.611 17 17 17c9.28 0 16.825-7.437 16.997-16.677L37 20h3c0 11.046-8.954 20-20 20S0 31.046 0 20C0 9.07 8.769.187 19.656.003L20 0Z"
        fill="#00347A"
      />
    </svg>
    <div
      class="c3 fi-loadingSpinner_text"
    >
      Loading data
    </div>
  </section>
</div>
`;
